### ISSUE TITLE
Node upgrades

### ISSUE THREAD

**Issue by mbforbes** [kubernetes/kubernetes#6079]
This is the roll-up issue for node upgrades. It contains the following issues:
- a minimum viable node upgrade mechanism for 1.0
- improvements upon the basic mechanism
- previous ones that discuss cluster upgrades and provisioning tools

Here are the details of our initial requirements for upgrade mechanisms:
- mechanism: **out-of-band** instead of on the master
- configuration: **salt** instead of another mechanism
- provider: **G*E** instead of a provider-agnostic solution
- node mgmt: **managed instance groups** instead of individual nodes
- must save storage: **best effort** instead of fully-supported
- language: **Bash** instead of Go
### 1.0 requirements

For 1.0, we are working on the tear down / recreate path:
- **Tear down / recreate:** Upgrading all components of a node together: kubelet, kube-proxy, Docker, and the underlying node image (the userspace and the kernel). This will be done by tearing down and creating nodes in the cluster with a newer supported version tuple. This is blocked by true static clustering&mdash;allowing nodes to dynamically join the cluster (#6087). This will be the "hammer" we can use for node upgrades and repairs: it is reliable as it starts with a fresh state, it covers all components of a node, but it is disruptive.
- [x] #6088 build path 2: provision new nodes to upgrade all components
- [x] #7914 test path 2
- [ ] #6084 Save storage: ensure volumes are retained
### Post-1.0

Post-1.0, we want to work on an in-place upgrade, and improve our existing mechanism to make it provider-agnostic, possibly run in a pod that manages the cluster, and other improvements listed below.
- **Binary upgrade:** Upgrading kubernetes binaries (kubelet, kube-proxy) independently of Docker and the node image. This is currently doable using `kube-push.sh` for some cloud providers. We need to ensure the cluster stays running in a good state after the upgrade, and support upgrading to official releases. This will cover most common upgrades, and future extensions to this mechanism (for example, to upgrade Docker) will allow for more thorough upgrades to be less disruptive, and help cover more use cases such as node repair.

The following are ordered by priority.
- [ ] #6099 build path 1: kubelet/kube-proxy in-place binary upgrade for 1.0.
- [ ] #7256 test path 1
- [ ] #6080 Node evacuation
- [ ] #6086 Only upgrade nodes whose versions can and should be upgraded
- [ ] #6095 Write in Go and hook into `pkg/cloudprovider`
- [ ] #6096 Make node upgrades work for all supported providers
- [ ] #6106 Put the node upgrade mechanism on the master
### Related work
- #2524 Full cluster upgrade discussion
- #3174 Looked at requirements for master upgrades (GCE)
- #4175 Put master data on a PD and reserved the master's IP (GCE)
- #3333 Node component upgrades
- #1573 Online kubelet upgrades
- #5044 Node rolling upgrades
- #4855 Full cluster upgrade discussion for 1.0; clarifies versioning
- #5511 Suggests `kubectl` command to do node rolling upgrades
- #5472 Proposal for `kubedeploy` to replace `kube-{up,down,push}.sh`
- #6075 Master upgrades roll-up

**Comment by mbforbes** [comment]
+cc @dchen1107 @davidopp @lavalamp @erictune @zmerlynn @alex-mohr

**Comment by mbforbes** [comment]
I'm leaving the project, so reassigning to... @zmerlynn! The other upgrade czar!

**Comment by aronchick** [comment]
+1 customer report of this today. 1.3?

**Comment by mikedanese** [comment]
@aronchick which part in particular? lot's of this is already done

**Comment by aronchick** [comment]
In particular, customers are very confused about the path for node upgrades. The items in "Post 1.0" are the ones they're looking for. Essentially:
- Node evacuation
- In-place k8s component upgrades
- Node rejoining

**Comment by davidopp** [comment]
Simple node evacuation is in 1.2 (`kubectl drain` / `kubectl uncordon`). @mml has a doc out with ideas for next steps (#17393), but we haven't had a chance to prioritize them relative to other efforts. 

In-place upgrade (kube-push.sh) definitely needs some love. #17397 is probably the best issue for tracking that.

I was looking into "node rejoining" the other day in response to a user question. I asked Yu-Ju "How does kubelet generate the node name? If I kill kubelet and restart it (or, equivalently, reboot the machine), will the kubelet join the cluster using the node name the previous kubelet was using, or a new one?" I think this is the core question for understanding node rejoining after a node upgrade. Her answer was:

> kubelet asks the cloud provider for the current node name. In the absence of this, kubelet uses the hostname (which can either be set by flags or from `uname -n). There is also an option in the kubelet configuration to set the node name.
>
> So, yes, kubelet will use the same name unless either of the condition listed above changes.
>
> Relevant code block:
> https://github.com/kubernetes/kubernetes/blob/575812787db7149ed458fda14544dc74a5e4f6e4/cmd/kubelet/app/server.go#L563

So I think node rejoining should "just work," both on cloud providers and bare metal.

**Comment by ravilr** [comment]
@davidopp @kubernetes/sig-cluster-ops 
with regards to node drain, when upgrading the nodes in a baremetal cluster in a rolling manner, the pods killed as part of drain gets replaced by new pods by the rc manager. is there any way to prevent the replaced pods from getting scheduled on still-to-be upgraded nodes.  Otherwise, the pods will be bounced around a lot of times till all/most of the nodes are finished upgrading. Some implicit predicate to prefer nodes running same version of kubelet as the master version?

**Comment by mikedanese** [comment]
The drain sets the node to unschedulable so this won't happen.

**Comment by ravilr** [comment]
@mikedanese 
I'm talking about having a way to avoid replaced pods getting binded to the other nodes in the cluster which are yet to be upgraded. Otherwise, when i get to upgrade those other nodes, the pods on it will be terminated again. this means every time a node is drained for version upgrade, it is possible that pods belonging to some rc/replicasets are continuously terminated and moved around nodes.

**Comment by davidopp** [comment]
Unfortunately this isn't possible right now. But coincidentally, we were recently discussing (see link directly above) exposing some of the things like Kubelet version as labels. If we did that, then it would be possible to do what you suggested (set a "preference" for the new kubelet version on all of your pods before doing the upgrade).

**Comment by kargakis** [comment]
What's the status of this issue?

**Comment by davidopp** [comment]
There is a huge constellation of node drain/update related issues floating around. I think that instead of playing whack-a-mole with them, someone when they have the time should write a comprehensive roadmap for where we want to go, with pointers to the relevant issues.

**Comment by roberthbailey** [comment]
This was tracking a feature improvement for kube-up, which is deprecated. Each installer does their own version of node upgrades, some of which are in-place.

/close
